ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,376 @@
+///   { IntegrityVariable, PublicInput, TraceColumns }
+fn get_iterable_len(
+    symbol_table: &SymbolTable,
+    iterable: &Iterable,
+) -> Result<usize, SemanticError> {
+    match iterable {
+        Iterable::Identifier(ident) => {
+            let ident_type = symbol_table.get_type(ident.name())?;
+            match ident_type {
+                IdentifierType::IntegrityVariable(var_type) => match var_type.value() {
+                    VariableType::Vector(vector) => Ok(vector.len()),
+                    _ => Err(SemanticError::InvalidListComprehension(
+                        "List comprehensions only allowed for vector variables.".to_string(),
+                    )),
+                },
+                IdentifierType::PublicInput(size) => Ok(*size),
+                IdentifierType::TraceColumns(trace_columns) => Ok(trace_columns.size()),
+                _ => Err(SemanticError::InvalidListComprehension(
+                    "IdentifierType {ident_type} not supported for list comprehensions".to_string(),
+                )),
+            }
+        }
+        Iterable::Range(range) | Iterable::Slice(_, range) => Ok(range.end() - range.start()),
+    }
+}
+
+/// Checks if the access index is valid. Returns an error if the access index is greater than
+/// the size of the vector.
+fn validate_access(i: usize, size: usize) -> Result<(), SemanticError> {
+    if i < size {
+        Ok(())
+    } else {
+        Err(SemanticError::InvalidListComprehension(
+            "Invalid access index".to_string(),
+        ))
+    }
+}
+
+/// Builds an [IterableContext] from a given list comprehension.
+fn build_iterable_context(lc: &ListComprehension) -> Result<IterableContext, SemanticError> {
+    let mut iterable_context = IterableContext::new();
+    for (member, iterable) in lc.context() {
+        if iterable_context
+            .insert(member.clone(), iterable.clone())
+            .is_some()
+        {
+            return Err(SemanticError::InvalidListComprehension(
+                "Duplicate member in list comprehension".to_string(),
+            ));
+        }
+    }
+    Ok(iterable_context)
+}
+
+/// Builds an [Expression] from a given identifier.
+///
+/// # Errors
+/// - Returns an error if the identifier is not of type in set:
+///  { IntegrityVariable, PublicInput, TraceColumns, RandomValuesBinding }
+/// - Returns an error if the access index is greater than the size of the vector.
+/// - Returns an error if the identifier is not a vector in the symbol table if it's a variable.
+fn build_ident_expression(
+    ident: &Identifier,
+    ident_type: &IdentifierType,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    match ident_type {
+        IdentifierType::TraceColumns(trace_columns) => {
+            validate_access(i, trace_columns.size())?;
+            Ok(Expression::NamedTraceAccess(NamedTraceAccess::new(
+                ident.clone(),
+                i,
+                CURRENT_ROW,
+            )))
+        }
```

Comment:
I know it would mean passing the symbol table, but it might be preferable to create an `IndexedTraceAccess` at this point. Since it's the IR we have all the information, and if we create a `NamedTraceAccess` we just have to convert it later.

Reply:
Makes sense. Done.